### ISSUE TITLE
Model implementation differs from what is described in the paper

### ISSUE THREAD

**Issue by sjosic** [ekzhang/fastseg#8]
Hi @ekzhang. Thank you for your implementation. I am trying to reimplement Lite R-ASPP MobileNet_v3 segmentation network in Keras by following your PyTorch implementation and description in the paper. The authors of the paper said that R-ASPP is a reduced design of the Atrous Spatial Pyramid Pooling module, which adopts only two branches consisting of a 1×1 convolution and a global-average pooling operation. Is there any particular reason why you used a total of 4 branches (aspp_conv1, aspp_conv2, aspp_conv3 and aspp_pool) in the case when Lite R-ASPP segmentation head is not used? In DeepLab_v3+ paper they used 5 branches is ASPP module and the simple yet effective decoder with only one branch from low level features while you are using 3 branches for your decoder. Also, the authors of the paper do not define whether it is necessary to use a decoder at all after the segmentation head. I think that they use Lite R-ASPP segmentation head as decoder itself. Your final output has output_stride=8, but authors attached Lite R-ASPP block at backbone with output_stride=16. Also, authors said that they used atrous convolution only in the last block of mobilenet v3 backbone but you are using atrous convolution in the last 4 blocks (blocks 3, 4, 5 and 6 with total 9 InvertedResidual and one last convolution block). I would be very grateful if you could clarify some of these differences. 

![DeepLab_v3_plus](https://user-images.githubusercontent.com/47183805/100750806-d725b780-33e6-11eb-96be-56aadf858eca.png)
![Lite R-ASPP](https://user-images.githubusercontent.com/47183805/100750807-d7be4e00-33e6-11eb-9f11-c314df2498f3.png)

**Comment by ekzhang** [comment]
Hello @sjosic, the LR-ASPP segmentation head is used when `use_aspp=False` is passed, which is the default. You may want to reread the `model/lraspp.py` source code.

As for the atrous convolution and decoder, you are correct that the authors are a bit vague. I implemented a segmentation network and was able to reach an mIOU approximately matching those presented in the paper. This is, to my knowledge, the only public MobileNetV3 implementation that reaches this mIOU. Unfortunately, there is no canonical implementation published by the authors, to my knowledge.

**Comment by sjosic** [comment]
Hello @ekzhang, thank you for your answer. Yes, you are right, when use_aspp=False is passed, which is the default, the LR-ASPP segmentation head is used. What about the situation when we want to use the ASPP segmentation head? My previous comment was about this situation (use_aspp=True). As I mentioned, the authors of the paper said that reduced ASPP is a reduced design of the Atrous Spatial Pyramid Pooling module, which adopts only two branches consisting of a 1×1 convolution and a global-average pooling operation. You used a total of 4 branches (aspp_conv1, aspp_conv2, aspp_conv3 and aspp_pool) in the case when use_aspp=True. In DeepLab_v3+ paper, they used 5 branches in the ASPP module. I think that your reduced atrous spatial pyramid pooling implementation (use_aspp=True) is something between the reduced ASPP (two branches) and DeepLab_v3+ version (5 branches). Regarding the decoder part, I downloaded mobilenetv3_large_cityscapes_trainfine checkpoint from https://github.com/tensorflow/models/blob/master/research/deeplab/g3doc/model_zoo.md and checked it in TensorBoard. They didn't use any decoder after the LR-ASPP segmentation head. I think they use LR-ASPP block as a decoder itself. Also in this checkpoint file, they used input shape=(1025, 2049, 3) and output_stride=32 (no atrous convolution) while you are using output_stride=8. Below are segmentation results using their large and small versions on two test images.

Large:
![mobilenetv3_large_cityscapes_trainfine_nyc](https://user-images.githubusercontent.com/47183805/101341666-b98e9d00-3881-11eb-9043-a903e5b2ab6f.png)
![mobilenetv3_large_cityscapes_trainfine_sf](https://user-images.githubusercontent.com/47183805/101341668-babfca00-3881-11eb-8196-fa1679769727.png)

Small:
![mobilenetv3_small_cityscapes_trainfine_nyc](https://user-images.githubusercontent.com/47183805/101341672-bb586080-3881-11eb-8485-5d2ec1848560.png)
![mobilenetv3_small_cityscapes_trainfine_sf](https://user-images.githubusercontent.com/47183805/101341675-bbf0f700-3881-11eb-9169-f7a1deec53be.png)

**Comment by ekzhang** [comment]
Thanks for the response, and happy new year. The `use_aspp` parameter was just for personal experimentation; it's not intended to exactly match either "ASPP" or "R-ASPP" as called by the authors, and I didn't release any pretrained weights with `use_aspp=True`.

As for the decoder part, thanks for finding the implementation in TensorFlow model zoo! It looks to me in their table that they have `output_stride = 8`, which matches my implementation. I haven't downloaded the actual model file, but yes it's possible that the decoder is slightly different -- I went off the paper when implementing this, so there were some missing details left to interpretation.

![image](https://user-images.githubusercontent.com/7550632/103968887-561ab800-512a-11eb-95f1-f4b9024259a1.png)

Hope this helps!

**Comment by sjosic** [comment]
Thank you for your response. I wish you all the best in this year. As you said, a lot of details are missing regarding the implementation in the paper itself. Also the TensorFlow implementation has a lot of unexplained details. For example there is `output_stride` and `decoder_output_stride` flag. `decoder_output_stride` is not mentioned in the paper. They simply said:
 
"When OS = 16, atrous convolution is applied in the last block of backbone. When OS = 32, no atrous convolution is used." 

When I checked in Tensorboard, the output of the backbone has OS = 32 (no atrous convolution was used at all in the last few backbone layers) but the output of the decoder part has OS = 8. I think that Eval OS from the table below is actual OS for the backbone. 

![Screenshot from 2021-02-20 14-32-37](https://user-images.githubusercontent.com/47183805/108597774-12eb7000-738b-11eb-96a5-86ea88fa7ba9.png)

At this moment I have Keras implementation which is very similar to the TensorFlow one, but I have problem to train this model on COCO dataset for example. I am getting black predictions. Model is good at predicting background pixels but it is not good at predicting foreground pixels. I think that problem is class imbalance in COCO dataset. I've tried different loss functions but without success.

**Comment by sjosic** [comment]
Hi @ekzhang In case you haven't seen it, the official PyTorch implementation for LR_ASPP MobileNet_V3 large model has been released: https://pytorch.org/vision/stable/models.html#semantic-segmentation. The pre-trained model has been trained on a subset of COCO train2017 (on the 20 categories that are present in the Pascal VOC dataset) and achieves 57.9 mIoU.

**Comment by ekzhang** [comment]
Thanks for letting me know! Very interesting to hear about.
